proxy/http2/HPACK.cc

```diff
@@ -877,38 +918,46 @@ hpack_encode_header_block(HpackIndexingTable &indexing_table, uint8_t *out_buf,
 
   MIMEFieldIter field_iter;
   for (MIMEField *field = hdr->iter_get_first(&field_iter); field != nullptr; field = hdr->iter_get_next(&field_iter)) {
-    HpackField field_type;
-    MIMEFieldWrapper header(field, hdr->m_heap, hdr->m_http->m_fields_impl);
     int name_len;
+    const char *name = field->name_get(&name_len);
     int value_len;
-    const char *name = header.name_get(&name_len);
-    header.value_get(&value_len);
+    const char *value = field->value_get(&value_len);
+
+    HpackHeaderField header{name, name_len, value, value_len};
+
     // Choose field representation (See RFC7541 7.1.3)
     // - Authorization header obviously should not be indexed
     // - Short Cookie header should not be indexed because of low entropy
-    if ((ptr_len_casecmp(name, name_len, MIME_FIELD_COOKIE, MIME_LEN_COOKIE) == 0 && value_len < 20) ||
-        (ptr_len_casecmp(name, name_len, MIME_FIELD_AUTHORIZATION, MIME_LEN_AUTHORIZATION) == 0)) {
+    HpackField field_type;
+    // TODO: replace `ptr_len_casecmp()` with `memcmp()`
+    if ((ptr_len_casecmp(header.name, header.name_len, HPACK_HDR_FIELD_COOKIE, HPACK_HDR_LEN_COOKIE) == 0 && value_len < 20) ||
+        (ptr_len_casecmp(header.name, header.name_len, HPACK_HDR_FIELD_AUTHORIZATION, HPACK_HDR_LEN_AUTHORIZATION) == 0)) {
       field_type = HpackField::NEVERINDEX_LITERAL;
     } else {
       field_type = HpackField::INDEXED_LITERAL;
     }
+
     const HpackLookupResult result = indexing_table.lookup(header);
+
+    if (result.match_type != HpackMatch::EXACT && field_type == HpackField::INDEXED_LITERAL) {
```

Comment:
I'm not sure what was unclear, but the intent was probably to delegate whole actual work. Now you can encode headers without updating indexing table, which should not happen. The old interface (with indexing_table param) ensures such an accident never happens.